apiserver/facades/client/application/application.go

```diff
@@ -1326,9 +1330,93 @@ func (api *APIBase) applicationSetCharm(
 		StorageConstraints: stateStorageConstraints,
 		EndpointBindings:   params.EndpointBindings,
 	}
+
+	// Disallow downgrading from a v2 charm to a v1 charm.
+	oldCharm, _, err := params.Application.Charm()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if charm.MetaFormat(oldCharm) >= charm.FormatV2 && charm.MetaFormat(newCharm) == charm.FormatV1 {
+		return errors.New("cannot downgrade from v2 charm format to v1")
+	}
+
+	// If upgrading from a pod-spec (v1) charm to sidecar (v2), override the
+	// application's series to what it would be for a fresh sidecar deploy.
+	oldSeries := params.Application.Series()
+	if oldSeries == "kubernetes" && charm.MetaFormat(newCharm) >= charm.FormatV2 &&
```

Comment:
you can have a v1 charm as kubernetes but have a v2 charm without sidecar containers. Remove len(containers) > 0

Reply:
In the meantime, I've reused the existing `core/charm.IsKubernetes` function (which incidentally has a comment about assumes).